### PR TITLE
Auto-skip: Basic support for WAIT

### PR DESCRIPTION
**Pull request by mikejholly**
This PR adds AutoSkip support for `WAIT...END` blocks.

### CODE REVIEW

File: inputgraph/inputgraph.go
Comment (vladaionescu) [inline inputgraph/inputgraph.go:201]: It might be helpful in case `WAIT` gets args in the future.

**Review by vladaionescu — COMMENTED** [review]
This looks good to me at first glance - I'll let @alexcb do the actual review.

File: inputgraph/inputgraph.go
Comment (alexcb) [reply inputgraph/inputgraph.go:201]: definitely useful -- I like the sorting of the Args -- that's actually something the existing code doesn't do a great job of.

a command like  `RUN --ssh --no-cache echo this last bits of the args matter`, will simply hash all args in the same order since we use a rather generic approach:

```
func (h *Hasher) HashCommand(cmd spec.Command) {
	dt, err := json.Marshal(cmd)
	if err != nil {
		panic(fmt.Sprintf("failed to hash command: %s", err)) // shouldn't happen
	}
	h.HashBytes(dt)
}
```

later on in the code it makes calls to `parseArgs` -- we could in theory figure out which of the Args represent flags (and can therefore be sorted), and which Args are part of the command can't be changed.

But this is all a rather small improvement to auto-skip in general, and could be pushed down the road.

**Review by alexcb — COMMENTED** [review]
looking good so far! seems like the WAIT/END blocks didn't need too much extra work.

ultimately I think most of the work is needed in extending the tests under `tests/autoskip/...` to improve our test coverage.

File: inputgraph/inputgraph.go
Comment (vladaionescu) [reply inputgraph/inputgraph.go:201]: Hmm.. I missed this part. Couldn't sorting lead to strange behavior? `RUN foo bar` being hashed the same as `RUN bar foo`?

Comment (alexcb) [reply inputgraph/inputgraph.go:201]: in the generic sense (like `RUN`, or `COPY`) yes; ~~however it's unlikely that the ordering of the future args (flags) for `WAIT` would matter.~~

actually the order does matter since we could have future flags like `WAIT --key value` rather than `WAIT --key=value`.

best to remove the `sort.Strings(waitStmt.Args)`.

If the user changes any ordering of flags in their Earthfile, I think it's fine for us to report it's a difference -- I don't think it's a use-case which we'll encounter frequently.

File: inputgraph/inputgraph.go
Comment (vladaionescu) [inline inputgraph/inputgraph.go:198]: This is in-lieu of the `HashCommand` call, right? I wonder if we should separate this in a similar function and co-locate the functions? I'm thinking that if we make changes to one, we might want to change the other too, and it's easier to spot that if the functions are next to each other.

**Review by vladaionescu — APPROVED** [review]
Looks good to me - just a minor remark. I would also wait for @alexcb's review, since he's more familiar with this code.

File: inputgraph/inputgraph.go
Comment (alexcb) [reply inputgraph/inputgraph.go:198]: this might be a good idea, because we already have `HashCommand`, `HashVersion`; however those commands just take a `spec.Command`, or `spec.Version` which are both leaf nodes of the ast.

WAIT on the other hand has references to other `spec.Command`, which means we have to walk the tree, and would need to call back into inputgraph package.

I honestly think it's fine keeping this as-is.


However, I think the current code might have an issue, as it's not clear how the following different blocks:

```
WAIT
  RUN echo foo
END
```

and

```
WAIT
END
RUN echo foo
```

will produce a different hash.

I think we would an additional:

`HashString("end")` or `HashInt(len(waitStmt.Body))`

File: util/buildkitskipper/hasher/hasher.go
Comment (alexcb) [inline util/buildkitskipper/hasher/hasher.go:36]: will `[]string{"a", "b"}` and `[]string{"a b"}` produce different hashes?

**Review by alexcb — CHANGES_REQUESTED** [review]
looking good; just a few questions regarding how we're encoding data to hash in a way that accounts for the number of elements, and if whitespace is significant or not.

File: inputgraph/inputgraph.go
Comment (mikejholly) [reply inputgraph/inputgraph.go]: Updated with a new function called `HashStatement` in favor of specific functions for `HashWait`, `HashIf`, etc.

File: util/buildkitskipper/hasher/hasher.go
Comment (alexcb) [reply util/buildkitskipper/hasher/hasher.go]: > They seem to produce the same hash.

We need to avoid cases like this -- any changes that the user makes to their Earthfile that could possibly propagate down to their container, should result in a different hash.

> Shall I use a different separator?

This will lead to the same issue. instead you'll need to come up with a more robust way to hash them to different values. That's why we were using `json.Marshal` in the `HashCommand` function.

File: util/buildkitskipper/hasher/hasher.go
Comment (alexcb) [inline util/buildkitskipper/hasher/hasher.go:49]: I think it's worthwhile that we include the datatype that we're hashing.

This will be a backwards-breaking change, but I think it's better to break things now rather than latter.

```suggestion
	h.h.Write([]byte(fmt.Sprintf("int:%d", i)))
```

and can you make a change to the HashString to include a `str:` prefix?

File: util/buildkitskipper/hasher/hasher.go
Comment (mikejholly) [reply util/buildkitskipper/hasher/hasher.go]: I was going to suggest a separator that cannot be used in the args, if one exists, but JSON is fine too. Updated.

File: util/buildkitskipper/hasher/hasher.go
Comment (mikejholly) [reply util/buildkitskipper/hasher/hasher.go]: Updated.

File: util/buildkitskipper/hasher/hasher.go
Comment (alexcb) [inline util/buildkitskipper/hasher/hasher.go:33]: This function seems very specific in requiring a single body, I think we should rename it to reflect that it's not a general use function, maybe just call it:

```suggestion
func (h *Hasher) HashWaitStatement(waitStmt spec.WaitStatement) {
```

honestly pushing down the `spec` into the hasher package feels like a wrong direction, which goes back to the original thread that @vladaionescu started.

The fact that we're passing in a `spec.*` yet only processing half of it in this package, and the other half in the `inputgraph` package feels like we have incorrectly broken this logic up into too many packages.

File: util/buildkitskipper/hasher/hasher.go
Comment (alexcb) [reply util/buildkitskipper/hasher/hasher.go]: awesome, thanks!

File: util/buildkitskipper/hasher/hasher.go
Comment (alexcb) [reply util/buildkitskipper/hasher/hasher.go:33]: ultimately I think the `HashCommand`, `HashVersion`, and now `HashWaitStatement` all belong in the `inputgraph` package.

perhaps this package could contain a helper function like `HashJSONMarshalled` which would take `v any`.

Comment (mikejholly) [reply util/buildkitskipper/hasher/hasher.go:33]: Yes, that's pretty much why I started with the WAIT-specific logic in `inputgraph.go`. It seems less desirable to have `Hasher` become aware of the specific `spec` types. Updated with suggestions.

File: util/buildkitskipper/hasher/hasher.go
Comment (alexcb) [inline util/buildkitskipper/hasher/hasher.go:32]: something seems odd with this, as we're going to end up hashing `str:int:123`.

I think it makes more sense to call `h.h.Write([]byte(fmt.Sprintf("int:%d", i)))` directly, so we don't end up with cases where we could easily end up with hashing collisions.

e.g. currently we have a collision between `HashInt(123)` and `HashString("int:123")`.

It would be a good idea to look over the API of this package to see if there's an potentials for collisions.

File: inputgraph/inputgraph.go
Comment (alexcb) [inline inputgraph/inputgraph.go:286]: this looks good, can we create a similar function for `hashCommand`

File: inputgraph/inputgraph.go
Comment (alexcb) [inline inputgraph/inputgraph.go:200]: we should set `v.SourceLocation = nil` here too

Comment (mikejholly) [reply inputgraph/inputgraph.go]: Updated.

File: inputgraph/inputgraph.go
Comment (mikejholly) [reply inputgraph/inputgraph.go:287]: Done.

File: util/buildkitskipper/hasher/hasher.go
Comment (mikejholly) [reply util/buildkitskipper/hasher/hasher.go]: Updated.

**Review by alexcb — COMMENTED** [review]
This is looking good! Thanks for adjusting the hasher package.

I think there's still some hashes in the tests that need to be updated since the entire hashing algorithm has changed.

**Comment by mikejholly** [comment]
> This is looking good! Thanks for adjusting the hasher package.
> 
> I think there's still some hashes in the tests that need to be updated since the entire hashing algorithm has changed.

I was confused by the fact that the tests are passing as is. I'll take a look at why.

https://cloud.earthly.dev/builds/6f60462e-6c2d-4d8c-8c52-96e4f646b399

**Comment by mikejholly** [comment]
@alexcb it looks like the existing tests are only looking for (or not looking for) the hash of a file rather than the auto-skip hash value.

**Review by alexcb — APPROVED** [review]

**Pull request merged**
